lib/AST/ASTContext.cpp

```diff
@@ -8407,6 +8413,48 @@ bool ASTContext::isEqualIgnoringChecked(QualType T1, QualType T2) const {
   }
 }
 
+// For the Checked C extension, compute whether a type is allowed to be an
+// argument or return type for a no-prototype function.   This computes the
+// set of allowed types described in Section 5.5 of the Checked C
+// specification.
+bool ASTContext::isNotAllowedForNoProtoTypeFunction(QualType QT) const {
+  if (const PointerType *PT = QT->getAs<PointerType>()) {
+    if (PT->isChecked())
+      return true;
+    if (PT->isFunctionPointerType())
+      return isNotAllowedForNoProtoTypeFunction(PT->getPointeeType());
```

Comment:
The `'is*` and`getAs` methods used in the conditional branches take care of desugaring and canonicalizing types if necessary.   The `getAs` method is a clang idiom specific to the Type hierarchy.